File: 15-3sum/15-3sum.java
class Solution {
 
    List<List<Integer>> res=  new ArrayList() ;
    public List<List<Integer>> threeSum(int[] nums) {
       
        Arrays.sort(nums) ;
        for(int i=0;i<nums.length;i++) {
            if(i==0 || nums[i-1]!=nums[i]) 
                two(i+1,nums.length-1,nums,0-nums[i]) ;
        }
        
        return res ;
      
    }
    public void two(int i,int j,int[] nums,int target ) {
       
        int a =nums[i-1] ; 
        while(i<j) {
            
           if(nums[i]+nums[j]>target) {
                j-- ;
            }
            else if(nums[i]+nums[j]<target) i++ ;
            else{
                List<Integer> array = new ArrayList() ;
                array.add(a);
                array.add(nums[i]) ;
                array.add(nums[j]) ;
                res.add(array);
                while(i<j && nums[i]==nums[i+1])i++ ;
                while(i<j && nums[j]==nums[j-1]) j-- ;
                i++; 
                j-- ;
            }
        }
    }
        
    
};

File: 15-3sum/15-3sum.cpp
class Solution {
public:
 vector<vector<int>> res;
    vector<vector<int>> threeSum(vector<int>& nums) {
    
        sort(nums.begin(),nums.end()) ;
        for(int i =0;i<nums.size();i++) {
            if(i==0 || nums[i-1]!=nums[i]){
                suyash(nums,i+1,nums.size()-1,0-nums[i]) ;
            }
        }
         
        return res ;
    }
    
    void suyash(vector<int> &nums,int a,int b,int target){
        int n =nums[a-1] ;
        
    while(a<b){
        
        if(nums[a]+ nums[b]>target) b-- ;
        else if(nums[a]+nums[b]<target) a++ ;
        
        else{
            vector<int> v; 
            v.push_back(n) ;
            v.push_back(nums[a]) ;
            v.push_back(nums[b]) ; 
            res.push_back(v) ;
            while(a<b && nums[a]==nums[a+1])a++ ;
            while(a<b && nums[b]==nums[b-1]) b-- ;
        a++; 
         b-- ;
        }
        
       
    }
        
        
    }

    
};